### ISSUE TITLE
Inline User Defined Types

### ISSUE THREAD

**Issue by mansellan** [WaynePhillipsEA/twinbasic#601]
**Is your feature request related to a problem?**
Given a type of the form:

```vb
Type EmployeeRecord
    ID As Integer
    Name As String * 20 
    Address As String * 30 
    Phone As Long 
    HireDate As Date 
End Type 
```

The type can be used as a `Sub` or `Function` parameter, or returned as a result of a `Function`. But sometimes such a type is only needed in a single call, and it seems wasteful to define a whole type (potentially polluting the global namespace) for a single use.

**Describe the solution you'd like**
It would be great if User Defined Types could be declared inline:

```vb
Public Sub UpdateCoordinates(Coordinates As Type(X As Double, Y As Double))
   ' Implementation skipped
End Sub

Public Function GetAddress(CustomerId As Integer) As Type(HouseNameOrNumber As String, Street As String, City As String, State As String)
   ' Implementation skipped
End Function
```

**Describe alternatives you've considered**
User defined types

**Additional context**
This would serve the exact same purpose as `ValueTuple` in VB.Net, except that members must always be named (a good thing IMO). User Defined Types seem extremely similar conceptually to `ValueTuple`, it's a shame VB.Net dispensed with them.

**Comment by WaynePhillipsEA** [comment]
So presumably, in order to use the returned UDT, we access the members via a `With` block?

```
With GetAddress(customerId)
   Debug.Print .Street
   Debug.Print .City
'... etc
End With
```

**Comment by mansellan** [comment]
So I think there's potentially a few options:
1. With blocks
2. Declare the same "shape" of UDT in the caller
3. A new inference capability? `Infer Address = GetAddress(customerId)

**Comment by WaynePhillipsEA** [comment]
Inference would be real easy to implement.  The hard part will be getting the community to agree on the syntax...

**Comment by WaynePhillipsEA** [comment]
For your `UpdateCoordinates` example, I assume you're thinking #600 will be used for passing the anonymous UDT members?

In my head, a lot of this does overlap with tuples.

**Comment by bclothier** [comment]
> In my head, a lot of this does overlap with tuples. 

I agree. However, I think we need to first address the more general problem of handling anonymous types & functions. We touched briefly on the anonymous functions in #436. How we solve this will then drive the solution here because conceptually, tuples are just anonymous structs that has names for elements.

**Comment by mansellan** [comment]
> For your `UpdateCoordinates` example, I assume you're thinking similar to #600 will be used for passing the anonymous UDT members?
> 
> In my head, a lot of this does overlap with tuples.

Hi, yes that was my thinking. To my mind, VBx pretty much already _has_ tuples in the form of UDTs, I don't see the point of reinventing them like VB.Net did, when we could just slightly expand the existing construct.

For inference, I agree the syntax needs discussion (and perhaps another issue)? To my mind a new keyword is appropriate this would be a language feature rather than an API. I proposed `Infer` as its nicely self-describing. I _think_ that the settings it could be used (basically instead of `Dim`, and perhaps only for locals) would mean it couldn't clash with user code, but I guess that would need thinking through.

>I agree. However, I think we need to first address the more general problem of handling anonymous types & functions. We touched briefly on the anonymous functions in #436. How we solve this will then drive the solution here because conceptually, tuples are just anonymous structs that has names for elements.

So these would be anonymous types, I guess I saw that as a slightly different concept from functions / delegates but agree there's overlap to be considered.

**Comment by mansellan** [comment]
I guess we'd also need to consider whether shape-matching would be needed (as suggested in point 2 above). It might be nice to be able to chain methods returning anonymous types together, even though the declarations are not in the same place.

I can see though that this could ramp up the complexity so maybe its one for later down the road.

**Comment by bclothier** [comment]
> So these would be anonymous types, I guess I saw that as a slightly different concept from functions / delegates but agree there's overlap to be considered.

Yes, we do want to make sure that the syntax we come up for handling anonymous types & methods and eventually for the tuples are all consistent and well-thought out. We don't want to go at each concept piecemeal and end up with a patchwork of syntax variants. Having a consistent syntax, IMO, is a _giant_ helper in driving adoption when new users are learning the language they don't want to have to learn 10 different syntaxes to do 10 different things. If they can learn 2 or 3 different syntaxes to do 10 different things, they're usually much more productive and can do a better job of predicting & learning new things in such language. 

Easy for me to state that; entirely another thing to actually come up with syntax that won't turn off a significant % of people, OTOH.... 😅

**Comment by EduardoVB** [comment]
IDK if this is included here in this proposal, but what I frequently have liked is to have the ability to define UDTs inside a procedure.

For example:

```
Private Sub Foo()
    Type POINTAPI
        X As Long
        Y As Long
    End Type
    Dim pt As POINTAPI

   GetCursorPos pt
   ' more code

End Sub
```

That combined with the possibility of defining API calls inside a procedure (mentioned somewhere else, I don't remember where) could allow to write procedures that don't depend on declarations from outside.

**Comment by mansellan** [comment]
I think you're thinking of [#406](https://github.com/WaynePhillipsEA/twinbasic/issues/406#issuecomment-917408850)
